File: backend/app/migrations.py
"""Pure, ordered document migrations. Storage owns backups and installation."""
from __future__ import annotations

from copy import deepcopy

from .errors import DomainError

SCHEMA_VERSION = 2
CAPABILITIES = ('project.revisions.v1',)
MAX_REVISION = 9007199254740991  # JSON/JavaScript safe integer, shared with Swift.


def version_1_to_2(document: dict) -> dict:
    result = deepcopy(document)
    result['schemaVersion'] = 2
    result['revision'] = 1
    result['requiredCapabilities'] = list(dict.fromkeys([
        *result.get('requiredCapabilities', []), 'project.revisions.v1']))
    return result


MIGRATIONS = {1: version_1_to_2}


def migrate_document(value: object) -> dict:
    if not isinstance(value, dict):
        raise DomainError('PROJECT_CORRUPT', 'The project document is damaged.')
    version = value.get('schemaVersion')
    if type(version) is not int or version < 1:
        raise DomainError('PROJECT_CORRUPT', 'The project schema version is invalid.')
    if version > SCHEMA_VERSION:
        raise DomainError('SCHEMA_UNSUPPORTED', 'Upgrade required: this project uses a newer format.', 409)
    capabilities = value.get('requiredCapabilities', [])
    if (not isinstance(capabilities, list) or len(capabilities) > 128
            or any(not isinstance(c, str) or not c or len(c) > 100 for c in capabilities)
            or len(set(capabilities)) != len(capabilities)):
        raise DomainError('PROJECT_CORRUPT', 'The project capability list is invalid.')
    if any(c not in CAPABILITIES for c in capabilities):
        raise DomainError('CAPABILITY_UNSUPPORTED', 'Upgrade required: this project needs unsupported capabilities.', 409)
    result = deepcopy(value)
    while result['schemaVersion'] < SCHEMA_VERSION:
        result = MIGRATIONS[result['schemaVersion']](result)
    if 'project.revisions.v1' not in result.get('requiredCapabilities', []):
        raise DomainError('PROJECT_CORRUPT', 'The project revision capability is missing.')
    if type(result.get('revision')) is not int or not 1 <= result['revision'] <= MAX_REVISION:
        raise DomainError('PROJECT_CORRUPT', 'The project revision is invalid.')
    return result


def runtime_capabilities() -> dict:
    return {'schemaVersion': SCHEMA_VERSION, 'requiredCapabilities': list(CAPABILITIES),
            'conditionalSave': True, 'conditionalExport': True, 'recoveryCopy': True,
            'exportRetry': True, 'storageBreakdown': True, 'exportFileCleanup': True,
            'projectCheckpoints': True, 'projectDuplicate': True, 'projectTrash': True, 'mediaJobs': True, 'proxyRepair': True, 'hdrToSdr': False}

File: frontend/src/project/migrations.ts
/** Pure migration registry. Services preserve the original before installation. */
export const SCHEMA_VERSION = 2;
export const CAPABILITIES = ['project.revisions.v1', 'media.hdr-to-sdr.v1'] as const;

export class ProjectError extends Error {
  constructor(
    public readonly code: string,
    message: string,
  ) {
    super(message);
    this.name = 'ProjectError';
  }
}

type Document = Record<string, unknown>;
export const migrations: Record<number, (document: Document) => Document> = {
  1: (document) => ({
    ...document,
    schemaVersion: 2,
    revision: 1,
    requiredCapabilities: [
      ...new Set([
        ...((document.requiredCapabilities as string[] | undefined) ?? []),
        'project.revisions.v1',
      ]),
    ],
  }),
};

export function migrateDocument(value: unknown): Document {
  if (!value || typeof value !== 'object' || Array.isArray(value))
    throw new ProjectError('PROJECT_CORRUPT', 'The project document is damaged.');
  const document = value as Document;
  const version = document.schemaVersion;
  if (!Number.isSafeInteger(version) || (version as number) < 1)
    throw new ProjectError('PROJECT_CORRUPT', 'The project schema version is invalid.');
  if ((version as number) > SCHEMA_VERSION)
    throw new ProjectError(
      'SCHEMA_UNSUPPORTED',
      'Upgrade required: this project uses a newer format.',
    );
  const capabilities =
    document.requiredCapabilities === undefined ? [] : document.requiredCapabilities;
  if (
    !Array.isArray(capabilities) ||
    capabilities.length > 128 ||
    capabilities.some((c: unknown) => typeof c !== 'string' || !c || c.length > 100) ||
    new Set(capabilities).size !== capabilities.length
  )
    throw new ProjectError('PROJECT_CORRUPT', 'The project capability list is invalid.');
  if (capabilities.some((c: string) => !(CAPABILITIES as readonly string[]).includes(c)))
    throw new ProjectError(
      'CAPABILITY_UNSUPPORTED',
      'Upgrade required: this project needs unsupported capabilities.',
    );
  let result = structuredClone(document);
  while ((result.schemaVersion as number) < SCHEMA_VERSION)
    result = migrations[result.schemaVersion as number](result);
  if (!(result.requiredCapabilities as string[] | undefined)?.includes('project.revisions.v1'))
    throw new ProjectError('PROJECT_CORRUPT', 'The project revision capability is missing.');
  return result;
}
